### PR TITLE
Mask passwords in while printing environment variables in our operators

### PR DESCRIPTION
**Pull request by scholzj**
### Type of change

- Bugfix

### Description

Our operators are at startup always printing the environment variables or other configurations. This is useful for analysing issues etc. But currently they are this way also printing variables with things such as passwords. This issue tires to detect them and mask them with `********`.

The new _masked_ output could look for example like this:

```
[2020-04-28 21:46:33,763] INFO  <Session     :71> [main        ] Using config:
        STRIMZI_TRUSTSTORE_LOCATION: /tmp/topic-operator/replication.truststore.p12
        STRIMZI_RESOURCE_LABELS: strimzi.io/cluster=my-cluster
        STRIMZI_KAFKA_BOOTSTRAP_SERVERS: my-cluster-kafka-bootstrap:9091
        STRIMZI_SSL_ENDPOINT_IDENTIFICATION_ALGORITHM: HTTPS
        STRIMZI_NAMESPACE: myproject
        STRIMZI_ZOOKEEPER_SESSION_TIMEOUT_MS: 20000
        STRIMZI_TOPICS_PATH: /strimzi/topics
        STRIMZI_FULL_RECONCILIATION_INTERVAL_MS: 90000
        STRIMZI_ZOOKEEPER_CONNECT: localhost:2181
        STRIMZI_TLS_ENABLED: true
        STRIMZI_KEYSTORE_PASSWORD: ********
        STRIMZI_TOPIC_METADATA_MAX_ATTEMPTS: 6
        STRIMZI_REASSIGN_VERIFY_INTERVAL_MS: 120000
        STRIMZI_KEYSTORE_LOCATION: /tmp/topic-operator/replication.keystore.p12
        TC_ZK_CONNECTION_TIMEOUT_MS: 20000
        STRIMZI_TRUSTSTORE_PASSWORD: ********
        STRIMZI_REASSIGN_THROTTLE: 9223372036854775807
```

### Checklist

- [x] Write tests
- [x] Make sure all tests pass
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally

### CODE REVIEW

**Comment by scholzj** [comment]
@tombentley Yeah, I know it is a bit hacky. But I went through the env vars and this is all I found. And I do not have a better way to deal with this right now.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java [hunk 1]
```diff
@@ -151,8 +151,8 @@ public KafkaMirrorMaker2AssemblyOperator(Vertx vertx, PlatformFeaturesAvailabili
                 .compose(i -> deploymentOperations.reconcile(namespace, mirrorMaker2Cluster.getName(), mirrorMaker2Cluster.generateDeployment(annotations, pfa.isOpenshift(), imagePullPolicy, imagePullSecrets)))
                 .compose(i -> deploymentOperations.scaleUp(namespace, mirrorMaker2Cluster.getName(), mirrorMaker2Cluster.getReplicas()))
                 .compose(i -> deploymentOperations.waitForObserved(namespace, mirrorMaker2Cluster.getName(), 1_000, operationTimeoutMs))
-                .compose(i -> deploymentOperations.readiness(namespace, mirrorMaker2Cluster.getName(), 1_000, operationTimeoutMs))
-                .compose(i -> reconcileConnectors(reconciliation, kafkaMirrorMaker2, mirrorMaker2Cluster, kafkaMirrorMaker2Status))
+                .compose(i -> mirrorMaker2Cluster.getReplicas() > 0 ? deploymentOperations.readiness(namespace, mirrorMaker2Cluster.getName(), 1_000, operationTimeoutMs) : Future.succeededFuture())
```
Comment (samuel-hawker) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java:154]: This seems like an unrelated change, I guess this was missed in the Connectors Exception fixes PR?

File: operator-common/src/main/java/io/strimzi/operator/common/Util.java [hunk 2]
```diff
@@ -223,4 +222,35 @@ private static File store(String prefix, String suffix, KeyStore trustStore, cha
...
+        for (Map.Entry<String, String> entry: env.entrySet()) {
+            sb.append("\t").append(entry.getKey()).append(": ").append(maskPassword(entry.getKey(), entry.getValue())).append("\n");
+        }
+
+        LOGGER.info("Using config:\n" + sb.toString());
+    }
+
+    /**
+     * Gets environment variable, checks if it contains a password and in case it does it masks the output. It expects
+     * environment variables with passwords to contain `PASSWORD` in their name.
```
Comment (samuel-hawker) [inline operator-common/src/main/java/io/strimzi/operator/common/Util.java:242]: Should we only be hiding passwords, might we want to hide apikeys or certificates as well?

Do let me know if I am wrong on this though...

Comment (samuel-hawker) [reply operator-common/src/main/java/io/strimzi/operator/common/Util.java:242]: If so I propose we expand the list of masks

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java (hunk 1 above)
Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java]: Good point, that was wrong commit. Fixed.

File: operator-common/src/main/java/io/strimzi/operator/common/Util.java (hunk 2 above)
Comment (scholzj) [reply operator-common/src/main/java/io/strimzi/operator/common/Util.java:242]: We do not pass any private (or even public) keys this way. So we actually don't know how would the environment variables look like. I can add some generic strings - e.g. "PRIVATE_KEY`, `API_KEY` - but not sure they would match in the future.

Comment (samuel-hawker) [reply operator-common/src/main/java/io/strimzi/operator/common/Util.java:242]: Makes sense to me, I was unsure if there were any and was just bringing it up in case there were any I was not aware of.

If you think it is unlikely that Strimzi will ever pass such values via Envars then I think you're right it is not needed, I do wonder though, if users are providing their own values for their own envars, e.g. in custom Connect images where they can configure extra envars for their own use cases might it make sense that we cover the hiding of such values by adding some extra generic strings like the ones you suggested above and briefly doc that these will be masked in the Envar list object that is used in the CRDs?

(Happy to not have it added though and only add it if a user requests such an enhancement)

Comment (scholzj) [reply operator-common/src/main/java/io/strimzi/operator/common/Util.java:242]: > If you think it is unlikely that Strimzi will ever pass such values via Envars then I think you're right it is not needed

I'm not sure I would call it unlikely. But until we have them, we do not have any name to filter against. So we just guess how our future self would name them. Or we should add them here hen we start using them.

> if users are providing their own values for their own envars, e.g. in custom Connect images where they can configure extra envars for their own use cases might it make sense that we cover the hiding of such values by adding some extra generic strings like the ones you suggested above and briefly doc that these will be masked in the Envar list object that is used in the CRDs?

This covers only the operators. Not things like Kafka Connect. Users can provide custom variables here as well, but there are no obvious plugable interfaces. So make any sense of passing something secret, they would need to modify the project quite significantly. Also, we again do not know the names, so it is hard to mask them.

So I would probably leave it out as it is.

Comment (samuel-hawker) [reply operator-common/src/main/java/io/strimzi/operator/common/Util.java:242]: Thanks for clarifying, happy to approve :)

Comment (scholzj) [reply operator-common/src/main/java/io/strimzi/operator/common/Util.java:242]: Thanks 👍
